recursive-methods.rb

```diff
@@ -1,36 +1,258 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
+
+# Write #1
+# factorial(n) Write a method factorial that accepts an integer parameter n and that uses recursion to compute and return the value of n factorial (also known as n!)
+# e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+
 def factorial(n)
-  puts "Not implemented."
+  if n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
+end
+
+if factorial(4) == 24
+  puts "Factorial alg: Right!"
+else
+  puts "Factorial alg: BUG"
 end
 
+#time comp: O(n) where n is the size of the digit. As the integer increases, the computational time increases at the same rate.
+#space comp: O(n) where n is the size of the number. Ad the number increases, the number of times the function is called grows at the same rate.
+
+# Write #2
+# reverse(s) Write a method reverse that accepts a string as a parameter and then returns the reverse of the string by reversing all letters and all words in the string.
+# e.g. reverse("hello world") will return "dlrow olleh"
+
 def reverse(s)
-  puts "Not implemented."
+  n = s.length - 1
+  if n == 0
+    return s
+  else
+    return s[-1] + reverse(s[0..(n-1)])
+  end
+  # if s[0] == s[-1]
+  #   return s
+  # else
+  #   return s[-1] + reverse_inplace(s[1..-2])+ s[0]
+  # end
 end
 
-def reverse_inplace(s)
-  puts "Not implemented."
+s = "hello world"
+if reverse(s) == "dlrow olleh"
+  puts "Reverse str alg: Right! #{s}"
+else
+  puts "Reverse str alg: BUG"
 end
 
+#time comp: O(n)
+#space comp: O(n) * O(n) = On^2 (it takes up O(n) space of memory and O(n) space to create a string each time)
+
+# Write #3
+# reverse_inplace(s) Write a method reverse_inplace that accepts a string as a parameter and then reverses the string in place using a recursive algorithm.
+# e.g. reverse("hello world") will convert the input string to "dlrow olleh"
+
+def reverse_inplace(s, i=0, j=(-1))
+
+  base_case = s.length / 2
+  if s[i] == s[j] || (i == base_case)
```

Comment:
Consider "toast" as s. In the first call, base_case will be computed to be 2. Then on line 59, s[i] i.e. s[0], which is 't' will equal s[j] i.e. s[4], which is also 't', and the method will return s. But "toast" won't be reversed.
To fix this, line 59 should be `if i >= j` or `if i == base_case`